feat: Enable precomputation of VID dispersal data

akonring opened the pull request:
## Description

Follows the approach described here: #479.

Introducing the subtrait `Precomputable` to a VID scheme.
This allows for re-using the polynomial commitments (from `commit_only_precompute`) when calling `disperse_precompute`.

Lots of code-duplication but was a bit reluctant to refactor and make pre-mature code optimizations since we might have more data that can potentially be pre-computed and keeping the disperse code as is (in one function) makes it easy to map tracing output to code (readability of benchmarks).

---

Preliminary testing for pre-computation approach:
`payload_size = 33MB, num_storage_nodes = 128, multiplicity = 8`

We get running times:
commit_only_precompute: 2.768s
**disperse_precompute: 2.580s**

(see test run below)
```
Espressos-MBP:jellyfish espressosystems$ cargo test --release --package jf-primitives --lib --features print-trace -- vid::advz::precomputable::tests::disperse_with_data_timer --exact --nocapture
   Compiling jf-primitives v0.4.0-pre.0 (/Users/espressosystems/projects/jellyfish/primitives)
    Finished release [optimized] target(s) in 17.71s
     Running unittests src/lib.rs (target/nix_rustc/release/deps/jf_primitives-a0dd78b380e04750)

running 1 test
Start:   KZG10::Setup with prover degree 512 and verifier degree 1
··Start:   Generating powers of G
··End:     Generating powers of G ..................................................4.297ms
End:     KZG10::Setup with prover degree 512 and verifier degree 1 .................7.846ms
Start:   batch commit 2115 polynomials
End:     batch commit 2115 polynomials .............................................2.768s
Start:   (PRECOMPUTE): VID disperse 33554432 payload bytes to 128 nodes
··Start:   encode payload into field elements
··End:     encode payload into field elements ......................................0ns
··Start:   field elements into polynomials (inverse FFT)
··End:     field elements into polynomials (inverse FFT) ...........................864.920ms
··Start:   compute all storage node evals for 2115 polynomials with 512 coefficients
··End:     compute all storage node evals for 2115 polynomials with 512 coefficients 871.298ms
··Start:   compute merkle root of all storage node evals
··End:     compute merkle root of all storage node evals ...........................237.428ms
··Start:   (PRECOMPUTE): compute 2115 KZG commitments
··End:     (PRECOMPUTE): compute 2115 KZG commitments ..............................28.417µs
··Start:   compute aggregate proofs for 128 storage nodes
····Start:   compute h_poly
····End:     compute h_poly ........................................................392.693ms
····Start:   gen eval proofs with parallel_factor 2 and num_points 1024
····End:     gen eval proofs with parallel_factor 2 and num_points 1024 ............170.916ms
··End:     compute aggregate proofs for 128 storage nodes ..........................571.669ms
··Start:   assemble shares for dispersal
··End:     assemble shares for dispersal ...........................................9.214ms
End:     (PRECOMPUTE): VID disperse 33554432 payload bytes to 128 nodes ............2.580s
test vid::advz::precomputable::tests::disperse_with_data_timer ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 113 filtered out; finished in 6.35s
```


---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

ggutoski commented on primitives/src/vid/advz/precomputable.rs:
```diff
@@ -0,0 +1,300 @@
+            polys.len()
+        ));
+        let common = Common {
+            poly_commits: data.poly_commits.clone(), /* UnivariateKzgPCS::batch_commit(&self.ck,
+                                                      * &polys).map_err(vid)?, */
```
```suggestion
            poly_commits: data.poly_commits.clone(),
```
please take out this comment

akonring replied:
fixed: 09e289a517a3af343c06017d2729df7412b25cdc

ggutoski commented on primitives/src/vid/advz/precomputable.rs:
```diff
@@ -0,0 +1,300 @@
+        let mut shares = Vec::with_capacity(code_word_size);
+        let mut evals = Vec::new();
+        let mut proofs = Vec::new();
+        for index in 0..code_word_size {
```
Looks like you removed a superfluous variable `i` relative to advz.rs. I like it. Please replicate this change back in advz.rs, too.

akonring replied:
Don't think anything has changed here. But maybe I've missed something

ggutoski replied:
it's now in `assemble_shares` and it's fixed

ggutoski reviewed (commented):
A few minor nits below.

Regarding this:
> Lots of code-duplication but was a bit reluctant to refactor and make pre-mature code optimizations since we might have more data that can potentially be pre-computed and keeping the disperse code as is (in one function) makes it easy to map tracing output to code (readability of benchmarks).

Not sure I agree here. It might be possible to write a few functions so that each version of disperse calls the same 5 functions, plus some bench label differences. But I'm not sure it's worth the time to bother. If you don't want to do it then please post a tech-debt github issue on reducing code duplication and link to it in a comment at the top of `precomputable.rs`.

akonring commented on primitives/src/vid/advz/precomputable.rs:
```diff
@@ -0,0 +1,300 @@
+// Copyright (c) 2023 Espresso Systems (espressosys.com)
```
fixed: 3ae78c01f51d4b0d25efafcfbb028725edba095e

akonring commented on primitives/src/vid/precomputable.rs:
```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2023 Espresso Systems (espressosys.com)
```
fixed: 3ae78c01f51d4b0d25efafcfbb028725edba095e

akonring commented on primitives/src/vid/advz/precomputable.rs:
```diff
@@ -0,0 +1,300 @@
+            UnivariateKzgPCS::batch_commit(&self.ck, &polys).map_err(vid)?;
+        let commit = Self::derive_commit(&poly_commits, payload.len(), self.num_storage_nodes);
+
+        commit.map(|c| (c, PrecomputeData { poly_commits }))
```
fixed: 492d462979e6e188996362ca01cc7794bbaf168a

akonring commented on primitives/src/vid/advz/precomputable.rs:
```diff
@@ -0,0 +1,300 @@
+        &self,
+        payload: B,
+        data: &Self::PrecomputeData,
+    ) -> crate::vid::VidResult<crate::vid::VidDisperse<Self>>
```
fixed: 8d0b26bb555ddd47f643957a3f73450549f4440c

akonring commented:
Thanks for the comments. I've addressed the code comments individually.

> ...But I'm not sure it's worth the time to bother. If you don't want to do it then please post a tech-debt github issue on reducing code duplication and link to it in a comment at the top of `precomputable.rs`.

I agree. Did some more refactoring of certain parts of the `disperse` code but I think we can do more (especially, when we find that the code is in a somewhat stable state). Let me know if we should still create an issue for this or you think it is tolerable, as is.

ggutoski commented on primitives/src/vid/advz.rs:
```diff
@@ -267,52 +263,22 @@ where
         let code_word_size = self.multiplicity * self.num_storage_nodes;
 
         // partition payload into polynomial coefficients
         // and count `elems_len` for later
```
```suggestion
```
I think this line of comment is no longer relevant

ggutoski commented on primitives/src/vid/advz.rs:
```diff
@@ -267,52 +263,22 @@ where
+        let all_storage_node_evals = self.evaluate_polys(&polys)?;
         end_timer!(all_storage_node_evals_timer);
 
         // vector commitment to polynomial evaluations
```
The next line can be deleted:
```rust
// TODO why do I need to compute the height of the merkle tree?
```

ggutoski commented on primitives/src/vid/advz.rs:
```diff
@@ -684,6 +680,45 @@ where
+    ///
+    /// Each share contains (for multiplicity m):
+    /// 1. (m * num_poly) evaluations.
+    /// 2. m aggregated KZG proofs.
```
```suggestion
    /// 2. a collection of m KZG proofs. TODO KZG aggregation https://github.com/EspressoSystems/jellyfish/issues/356
```

ggutoski reviewed (commented):
The refactoring's an improvement---thank you!

I noticed a few obsolete comments. Please bundle those in when you push commits to fix CI.
